@lingosta/app/providers/useUser.tsx

```diff
@@ -0,0 +1,87 @@
+"use client";
+import { useState, useEffect, useContext, createContext } from 'react';
+import { account } from '../utils/appwrite';
+import { AppwriteException } from 'appwrite';
+import { useRouter } from 'next/navigation';
+
+import toast from 'react-hot-toast';
```

Comment:
Let's not change anything for now, but I have some concerns about having toasts inside of the login functionality.

I'm keen on decoupling UI-behavior from the provider. The provider should be responsible for handling the user logic. Any errors, notifications, based off this can be deferred to whatever called it. And whatever called it can wrap it in a try-catch block.

This isn't necessarily wrong, but the code can become a bit hard to maintain if UI is tightly-coupled with other logic. You can imagine other cases where we might want to do user-related stuff and now it's stuck with the toasts that are built-in.

For our smaller app, I think it should be okay. But for a larger application that can grow more complex, this is something to be wary of.

Reply:
Also makes it less flexible for us to incorporate other stuff that we might want the UI to do. For example, if we want somethign else to handle the UI behavior on top of what's alread ybaked in.

Reply:
good point, I'll note that for future logic